### PR TITLE
comment template let

### PR DESCRIPTION
**Pull request by lkarthee**
comment some wip bindings to enable tests in tko.bind.

### CODE REVIEW

File: src/templateRewriting.js [hunk 1]
```diff
@@ -1,66 +1,84 @@
...
+import {
+  parseObjectLiteral
+} from 'tko.provider';
+
+
+var memoizeDataBindingAttributeSyntaxRegex = /(<([a-z]+\d*)(?:\s+(?!data-bind\s*=\s*)[a-z0-9\-]+(?:=(?:\"[^\"]*\"|\'[^\']*\'|[^>]*))?)*\s+)data-bind\s*=\s*(["'])([\s\S]*?)\3/gi;
+var memoizeVirtualContainerBindingSyntaxRegex = /<!--\s*ko\b\s*([\s\S]*?)\s*-->/g;
+
+function validateDataBindValuesForRewriting(keyValueArray) {
+  /*var allValidators = ko.expressionRewriting.bindingRewriteValidators;
```
Comment (brianmhunt) [inline src/templateRewriting.js]: Do we still need template rewriting at all?  I'm just not clear on what it does ... Or if there is now a much simpler way in ES5.

File: src/templateSources.js [hunk 2]
```diff
@@ -119,6 +119,6 @@ anonymousTemplate.prototype.text = function(/* valueToWrite */) {
     }
 };
 
-ko.exportSymbol('templateSources', ko.templateSources);
-ko.exportSymbol('templateSources.domElement', ko.templateSources.domElement);
-ko.exportSymbol('templateSources.anonymousTemplate', ko.templateSources.anonymousTemplate);
+//ko.exportSymbol('templateSources', ko.templateSources);
```
Comment (brianmhunt) [inline src/templateSources.js]: We can just delete these exportSymbols - it's trivial to grep the original source for exports.

**Comment by brianmhunt** [comment]
There's a lot in the template binding code that I haven't taken the time to understand so please bear with me as I mull this 😀

File: index.js [hunk 3]
```diff
@@ -1,10 +1,16 @@
...
+import {foreach} from './src/foreach'
+import {template} from './src/templating'
 
 export var bindings = {
     'if': $if,
     'with': $with,
     ifnot: ifnot, unless: ifnot,
-    'let': letBinding,
-    template: template,
+    foreach: foreach,
```
Comment (brianmhunt) [inline index.js:11]: Relevant note:  I am hoping to switch the default foreach binding provider to https://github.com/brianmhunt/knockout-fast-foreach (which will become tko.binding.foreach, I guess).  The old binding behaviour will still be available with `template: { foreach: $data }`.

File: spec/ifBehaviors.js [hunk 4]
```diff
@@ -1,31 +1,64 @@
...
         testNode.innerHTML = "<div data-bind='if: someItem'><span data-bind='text: someItem.nonExistentChildProp'></span></div>";
         expect(testNode.childNodes[0].childNodes.length).toEqual(1);
-        ko.applyBindings({ someItem: null }, testNode);
+        applyBindings({ someItem: null }, testNode);
         expect(testNode.childNodes[0].childNodes.length).toEqual(0);
     });
 
-    it('Should leave descendant nodes in the document (and bind them) if the value is truthy, independently of the active template engine', function() {
-        this.after(function() { ko.setTemplateEngine(new ko.nativeTemplateEngine()); });
+    /*it('Should leave descendant nodes in the document (and bind them) if the value is truthy, independently of the active template engine', function() {
```
Comment (brianmhunt) [inline spec/ifBehaviors.js]: The canonical way to [disable tests in Jasmine 1.3](http://jasmine.github.io/1.3/introduction.html#section-Disabling_Specs_and_Suites) is with `xit` or `xdescribe`, I believe.

We should also note in the conversation why we have disabled a test.

File: spec/ifnotBehaviors.js [hunk 5]
```diff
@@ -1,31 +1,64 @@
...
         testNode.innerHTML = "<div data-bind='ifnot: condition'><span data-bind='text: someItem.nonExistentChildProp'></span></div>";
         expect(testNode.childNodes[0].childNodes.length).toEqual(1);
-        ko.applyBindings({ someItem: null, condition: true }, testNode);
+        applyBindings({ someItem: null, condition: true }, testNode);
         expect(testNode.childNodes[0].childNodes.length).toEqual(0);
     });
 
-    it('Should leave descendant nodes in the document (and bind them) if the value is falsey, independently of the active template engine', function() {
-        this.after(function() { ko.setTemplateEngine(new ko.nativeTemplateEngine()); });
+    /*it('Should leave descendant nodes in the document (and bind them) if the value is falsey, independently of the active template engine', function() {
```
Comment (brianmhunt) [inline spec/ifnotBehaviors.js]: Same deal re `xit` :smile:

File: spec/templatingBehaviors.js [hunk 6]
```diff
@@ -40,76 +69,92 @@ var dummyTemplateEngine = function (templates) {
         var templateOptions = options.templateOptions; // Have templateOptions in scope to support [js:templateOptions.foo] syntax
 
         var result;
-        with (bindingContext) {
```
Comment (brianmhunt) [inline spec/templatingBehaviors.js]: We've got to get `with` references out at some point.  Being generally an anti-pattern they tend to be confusing and also unoptimized in the browser.  We may need some alternative `lookup` here.

... if we need template rewriting at all ... what does it do? lol

File: spec/withBehaviors.js [hunk 7]
```diff
@@ -177,20 +210,20 @@ describe('Binding: With', function() {
         expect(testNode).toContainHtml("hello <!-- ko with: topitem --><!-- /ko -->");
     });
 
-    it('Should provide access to an observable viewModel through $rawData', function() {
+    /*it('Should provide access to an observable viewModel through $rawData', function() {
```
Comment (brianmhunt) [inline spec/withBehaviors.js]: `xit`

File: src/foreach.js [hunk 8]
```diff
@@ -1,40 +1,55 @@
+
+import {
```
Comment (brianmhunt) [inline src/foreach.js:2]: I commented elsewhere, but the `foreach` binding itself may be replaced by knockout-fast-foreach ... but we can still export / make it available.

File: src/templating.js [hunk 9]
```diff
@@ -1,308 +1,349 @@
...
+}
+
+function keyValueArrayContainsKey(keyValueArray, key) {
+    for (var i = 0; i < keyValueArray.length; i++)
+        if (keyValueArray[i]['key'] == key)
+            return true;
+    return false;
+}
+
+export function renderTemplate(template, dataOrBindingContext, options, targetNodeOrNodeArray, renderMode) {
```
Comment (brianmhunt) [inline src/templating.js:172]: This _feels_ somewhat overly complex, and that may just me being naive, but I feel like if we drop IE<8 support this gets simpler ... but again ... probably me being naive. :grin:

... and in any case, that's a future target; right now we want to stay even-keel until we get an ES6 build that passes all the tests and operates as a drop-in for Knockout.

**Comment by brianmhunt** [comment]
Great PR -- are you ok looking through my comments @lkarthee ?  Feel free to update the PR with any changes, or let me know and I will start merging.

Having always gone the easy route (jQuery), I am particularly unfamiliar & uncomfortable with a lot of what the template binding is doing.

One thing I want to keep in mind is that there are a lot of template renderers that would make sense here e.g. React, jsx, virtual-dom, and others.  By the time we hit something we can call Knockout 4.0 it would be good to close off a few issues in this respect e.g. knockout/knockout#1650  (yes it's closed, but the demand is still there), by making sure that alternative template designs plug in without having to rewrite from scratch.

File: spec/ifBehaviors.js (hunk 4 above)
Comment (lkarthee) [reply spec/ifBehaviors.js]: I forgot to uncomment. Will remove the comments.

**Pull request merged**